### PR TITLE
Add conversion of schema into base64

### LINKED ISSUE #27: Add some commands to convert the schema into base64
**Task description**

This command will make it easier to have a nice flow at presentations/workshops to explain where the magic `rawSchemaBase64` string comes from that is used in the front-end code.

```
cargo concordium schema-base64 --schema ./schema.bin --out ./my/path/
cargo concordium schema-base64 --module ./foo.wasm.v1 --out ./my/path/

cargo concordium build --schema-base64-out ./my/path/
```

The long-term plan is to support/promote `json-schema` instead which could give a better developer experience since they are human-readable and can be used for debugging purposes.

### PR DESCRIPTION
**Pull request by DOBEN**
## Purpose

closes #27 

- Add `schema-base64` command to convert a given schema to base64 format.
- Add `--schema-base64-out` option to `cargo concordium build` to optionally output the schema in base64 format.
-  Provding the value `-`  to the  `--schema-base64-out` flag will log the bsae64 string to the console instead of writing it into a file.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: cargo-concordium/src/build.rs [hunk 1]
```diff
@@ -410,7 +413,7 @@ fn write_schema_json(
     out_path.push(file_name);
 
     println!(
-        "   Writing schema for {} to {}.",
+        "   Writing json schema for {} to {}.",
```
Comment (abizjak) [inline cargo-concordium/src/build.rs:416]:
```suggestion
        "   Writing JSON schema for {} to {}.",
```

File: cargo-concordium/src/build.rs [hunk 2]
```diff
@@ -419,6 +422,46 @@ fn write_schema_json(
...
 }
 
+/// Write the provided schema in its base64 representation to a file inside the
+/// `root` directory if `Some(root)` is given.
+/// Print the provided schema in its base64 representation to the console if
+/// `schema_base64_log` is true.
+pub fn write_schema_base64(
+    root: Option<PathBuf>,
+    schema: &VersionedModuleSchema,
+    schema_base64_log: bool,
```
Comment (abizjak) [inline cargo-concordium/src/build.rs:432]: Why is this necessary? I think we should just output to `stdout` if `root = None`.

File: cargo-concordium/src/build.rs [hunk 3]
```diff
@@ -419,6 +422,46 @@ fn write_schema_json(
...
+    if let Some(root) = root {
+        let mut out_path = root;
+
+        let manifest = Manifest::from_path("Cargo.toml").context("Could not read Cargo.toml.")?;
+
+        let package = manifest
+            .package
+            .context("Manifest needs to specify [package]")?;
+
+        out_path.push(to_snake_case(package.name) + "_schema.b64");
```
Comment (abizjak) [inline cargo-concordium/src/build.rs:454]: I think the logic here should be the same as for `--schema`, namely that they provide a **file name**. Since only one file will be written.

File: cargo-concordium/src/main.rs [hunk 4]
```diff
@@ -132,7 +132,60 @@ enum Command {
...
+                 with this command."
+    )]
+    SchemaBase64 {
+        #[structopt(
+            name = "out",
+            long = "out",
+            short = "o",
+            default_value = ".",
+            help = "Writes the converted bsae64 representation of the schema to a file at the \
+                    specified location. Directory path must exist. (expected input: `./my/path/`)."
```
Comment (abizjak) [inline cargo-concordium/src/main.rs:149]: Remove default, and change this to a file.

File: cargo-concordium/src/main.rs [hunk 5]
```diff
@@ -132,7 +132,60 @@ enum Command {
...
+                    `./my/path/module.wasm.v1`)."
+        )]
+        module_path:       Option<PathBuf>,
+        #[structopt(
+            name = "log",
+            long = "log",
+            short = "l",
+            help = "Prints the base64 schema string to console."
+        )]
+        schema_base64_log: bool,
```
Comment (abizjak) [inline cargo-concordium/src/main.rs:187]: We should not need this. We should output to stdout if the file is not given.

**Review by limemloh — APPROVED** [review]
Nothing to add, which is not already a comment.
Looks good to me.

File: cargo-concordium/src/main.rs [hunk 6]
```diff
@@ -405,67 +463,50 @@ pub fn main() -> anyhow::Result<()> {
...
+        Command::SchemaBase64 {
+            out,
+            module_path,
+            schema_path,
+            wasm_version,
+        } => {
+            let schema = get_schema(module_path, schema_path, wasm_version)
+                .context("Could not get schema.")?;
+
+            if out == PathBuf::from("-") {
```
Comment (abizjak) [inline cargo-concordium/src/main.rs:480]:
```suggestion
            if out.as_path() == Path::new("-") {
```

File: cargo-concordium/src/main.rs [hunk 7]
```diff
@@ -405,67 +463,50 @@ pub fn main() -> anyhow::Result<()> {
...
+                write_schema_base64(None, &schema).context("Could not log base64 schema.")?;
```
Comment (abizjak) [inline cargo-concordium/src/main.rs:481]:
```suggestion
                write_schema_base64(None, &schema).context("Could not print base64 schema.")?;
```

File: cargo-concordium/src/main.rs [hunk 8]
```diff
@@ -526,7 +567,24 @@ pub fn main() -> anyhow::Result<()> {
                          (expected input `./my/path/`)."
                     );
                     write_json_schema(&schema_json_out, module_schema)
-                        .context("Could not write schema file.")?;
+                        .context("Could not write JSON schema files.")?;
+                }
+                if let Some(schema_base64_out) = schema_base64_out {
+                    if schema_base64_out == PathBuf::from("-") {
```
Comment (abizjak) [inline cargo-concordium/src/main.rs:573]:
```suggestion
                    if schema_base64_out.as_path() == Path::new("-") {
```

File: cargo-concordium/src/main.rs [hunk 9]
```diff
@@ -132,7 +132,56 @@ enum Command {
...
+    )]
+    SchemaBase64 {
+        #[structopt(
+            name = "out",
+            long = "out",
+            short = "o",
+            default_value = "-",
+            help = "Path and filename to a file to write the converted base64 representation to \
+                    or use the default value `-` to print the base64 schema to the console. \
+                    Directory path must exist. (expected input: `./my/path/base64_schema.b64` or \
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply cargo-concordium/src/main.rs]: Renamed it to:

```Path and filename to write the converted base64 representation to or use the default value `-` to print the base64 schema to the console. The path has to exist while the file will be created. (expected input: `./my/path/base64_schema.b64` or `-`).```

File: cargo-concordium/src/main.rs [hunk 10]
```diff
@@ -144,42 +193,51 @@ enum Command {
...
                     directory."
         )]
-        schema_json_out: Option<PathBuf>,
+        schema_json_out:   Option<PathBuf>,
+        #[structopt(
+            name = "schema-base64-out",
+            long = "schema-base64-out",
+            short = "b",
+            help = "Builds the contract schema and writes it in base64 format to the specified \
+                    directory/filename or prints the base64 schema to the console if the value \
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply cargo-concordium/src/main.rs]: renamed it to:

```
Builds the contract schema and writes it in base64 format to file at specified location or prints the base64 
schema to the console if the value `-` is used. The path has to exist while the file will be created. 
(expected input: `./my/path/base64_schema.b64` or `-`).
```

**Review by abizjak — APPROVED** [review]
